### ISSUE TITLE
[Architecture] Extract Backward Operations from Tape Module

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2633]
## Problem

The 879-line tape module needs reorganization to separate backward operations from core tape logic (detailed in #2609).

## Relationship to Other Issues

This implements the reorganization described in #2609 (Reorganize Tape Module Structure).

## Implementation Plan

1. Create `shared/autograd/backward_ops.mojo`
2. Move all `_backward_*` methods (15+ functions) to new file
3. Update imports in `tape.mojo`
4. Implement dispatch table from #2586
5. Test all backward operations

## File Structure

```
shared/autograd/
├── tape.mojo (300-400 lines)
│   └── Core tape logic, dispatch
└── backward_ops.mojo (500 lines)
    └── All backward implementations
```

## Benefits

- Clearer separation of concerns
- Easier navigation
- Better for code review
- Easier to add new operations

## Related Issues

- Implements #2609 (Reorganize Tape Module)
- Works with #2586 (Backward pass dispatch table)

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

I have gathered all the necessary information. Let me present the comprehensive implementation plan directly:

---

# Implementation Plan: Extract Backward Operations from Tape Module

**Issue**: #2633 - [Architecture] Extract Backward Operations from Tape Module  
**Related Issues**: #2609 (Reorganize Tape Module - CLOSED), #2586 (Backward pass dispatch table - CLOSED)

---

## 1. Summary

This plan extracts all backward operation implementations from the 792-line `tape.mojo` file into a dedicated `backward_ops.mojo` module. The extraction will:

1. Move 10 `_backward_*_by_idx` methods (~120 lines) to `shared/autograd/backward_ops.mojo`
2. Keep the dispatch mechanism in `tape.mojo` (current if/elif chain)
3. Update imports to use the new module structure
4. Maintain backward compatibility for all existing code

**Note**: The dispatch table optimization from #2586 is deferred due to Mojo v0.25.7 lacking Dict in stdlib.

---

## 2. Current State Analysis

### File: `/home/mvillmow/ml-odyssey/shared/autograd/tape.mojo` (792 lines)

| Section | Lines | Description |
|---------|-------|-------------|
| Module docstring | 1-48 | Comprehensive documentation |
| Imports | 50-69 | Backward functions from shared.core |
| Operation aliases | 72-89 | 17 operation type constants (OP_ADD, etc.) |
| SavedTensors struct | 91-166 | Container for tensors saved during forward pass |
| TapeNode struct | 168-259 | Represents single operation in computation graph |
| VariableRegistry struct | 261-378 | Maps variable IDs to gradient tensors |
| GradientTape struct | 380-604 | Core tape logic with dispatch |
| **Backward implementations** | **611-730** | **10 _backward_*_by_idx methods (to extract)** |
| NoGradContext | 733-792 | Context manager stub |

### Backward Methods to Extract (10 methods, ~120 lines)

| Method | Lines | Type | Tensors Saved |
|--------|-------|------|---------------|
| `_backward_add_by_idx` | 611-621 | Binary arithmetic | 2 |
| `_backward_subtract_by_idx` | 623-635 | Binary arithmetic | 2 |
| `_backward_multiply_by_idx` | 637-649 | Binary arithmetic | 2 |
| `_backward_divide_by_idx` | 651-663 | Binary arithmetic | 2 |
| `_backward_sum_by_idx` | 665-675 | Reduction | 1 + scalar |
| `_backward_mean_by_idx` | 677-687 | Reduction | 1 + scalar |
| `_backward_matmul_by_idx` | 689-701 | Matrix | 2 |
| `_backward_relu_by_idx` | 703-710 | Activation | 1 |
| `_backward_sigmoid_by_idx` | 712-721 | Activation | 1 |
| `_backward_tanh_by_idx` | 723-730 | Activation | 1 |

### Current `/home/mvillmow/ml-odyssey/shared/autograd/backward_ops.mojo` (30 lines)

Currently a stub that re-exports GradientTape. Will be completely rewritten.

---

## 3. Step-by-Step Implementation Tasks

### Task 1: Design the BackwardOps Interface

**Challenge**: Mojo methods cannot be extracted to another file and remain struct methods.

**Solution**: Use standalone functions that receive tape components as parameters:

```mojo
fn backward_add(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for addition by node index."""
    ...
```

**Rationale**:
- Avoids circular imports (backward_ops doesn't need GradientTape)
- Pass only what's needed: `nodes` (read-only) and `registry` (mutable)
- Clean separation of concerns

---

### Task 2: Create the New `backward_ops.mojo` Module

**File**: `/home/mvillmow/ml-odyssey/shared/autograd/backward_ops.mojo`

```mojo
"""Backward operation implementations for automatic differentiation.

This module provides backward pass implementations for all operation types
recorded in the GradientTape. Each function computes gradients using the
chain rule and stores them in the tape's registry.

Operation Categories:
- Binary arithmetic: add, subtract, multiply, divide
- Matrix operations: matmul
- Reduction operations: sum, mean
- Activation functions: relu, sigmoid, tanh

Design Note:
    Backward operations are implemented as standalone functions that receive
    tape components (nodes, registry) rather than the full GradientTape.
    This avoids circular imports and allows clear separation of concerns.
"""

from shared.core.extensor import ExTensor
from shared.core.arithmetic import (
    add_backward, subtract_backward, multiply_backward, divide_backward,
)
from shared.core.reduction import sum_backward, mean_backward
from shared.core.matrix import matmul_backward
from shared.core.activation import relu_backward, sigmoid_backward, tanh_backward

# Import types from tape (no GradientTape - avoids circular import)
from shared.autograd.tape import TapeNode, VariableRegistry

# ============================================================================
# Binary Arithmetic Operations
# ============================================================================

fn backward_add(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for element-wise addition."""
    if len(nodes[idx].saved.tensors) < 2:
        return
    var a = nodes[idx].saved.tensors[0]
    var b = nodes[idx].saved.tensors[1]
    var result = add_backward(grad_output, a, b)
    if len(nodes[idx].input_ids) >= 1:
        registry.set_grad(nodes[idx].input_ids[0], result.grad_a)
    if len(nodes[idx].input_ids) >= 2:
        registry.set_grad(nodes[idx].input_ids[1], result.grad_b)

fn backward_subtract(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for element-wise subtraction."""
    # Similar pattern...

fn backward_multiply(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for element-wise multiplication."""
    # Similar pattern...

fn backward_divide(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for element-wise division."""
    # Similar pattern...

# ============================================================================
# Reduction Operations
# ============================================================================

fn backward_sum(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for sum reduction."""
    if len(nodes[idx].saved.tensors) < 1:
        return
    var x = nodes[idx].saved.tensors[0]
    var axis = -1
    if len(nodes[idx].saved.scalars) >= 1:
        axis = Int(nodes[idx].saved.scalars[0])
    var grad_input = sum_backward(grad_output, x, axis)
    if len(nodes[idx].input_ids) >= 1:
        registry.set_grad(nodes[idx].input_ids[0], grad_input)

fn backward_mean(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for mean reduction."""
    # Similar pattern...

# ============================================================================
# Matrix Operations
# ============================================================================

fn backward_matmul(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for matrix multiplication."""
    # Similar pattern to binary ops...

# ============================================================================
# Activation Functions
# ============================================================================

fn backward_relu(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for ReLU activation."""
    if len(nodes[idx].saved.tensors) < 1:
        return
    var x = nodes[idx].saved.tensors[0]
    var grad_input = relu_backward(grad_output, x)
    if len(nodes[idx].input_ids) >= 1:
        registry.set_grad(nodes[idx].input_ids[0], grad_input)

fn backward_sigmoid(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for sigmoid activation."""
    # Similar pattern...

fn backward_tanh(
    nodes: List[TapeNode],
    mut registry: VariableRegistry,
    idx: Int,
    grad_output: ExTensor,
) raises:
    """Backward pass for tanh activation."""
    # Similar pattern...
```

---

### Task 3: Update `tape.mojo` Dispatch Mechanism

**Changes to `/home/mvillmow/ml-odyssey/shared/autograd/tape.mojo`**:

1. **Add import** for backward_ops functions (after line 69):

```mojo
from shared.autograd.backward_ops import (
    backward_add,
    backward_subtract,
    backward_multiply,
    backward_divide,
    backward_sum,
    backward_mean,
    backward_matmul,
    backward_relu,
    backward_sigmoid,
    backward_tanh,
)
```

2. **Update `_dispatch_backward_op`** method (lines 506-546):

```mojo
fn _dispatch_backward_op(
    mut self, op_type: String, node_idx: Int, grad_output: ExTensor
) raises:
    """Dispatch backward pass computation for the given operation type."""
    if op_type == OP_ADD:
        backward_add(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_SUBTRACT:
        backward_subtract(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_MULTIPLY:
        backward_multiply(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_DIVIDE:
        backward_divide(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_SUM:
        backward_sum(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_MEAN:
        backward_mean(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_MATMUL:
        backward_matmul(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_RELU:
        backward_relu(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_SIGMOID:
        backward_sigmoid(self.nodes, self.registry, node_idx, grad_output)
    elif op_type == OP_TANH:
        backward_tanh(self.nodes, self.registry, node_idx, grad_output)
    else:
        raise Error("Unsupported operation type: " + op_type)
```

3. **Remove all `_backward_*_by_idx` methods** (lines 611-730)

---

### Task 4: Handle Circular Import Prevention

**Key Insight**: `backward_ops.mojo` imports from `tape.mojo` (TapeNode, VariableRegistry), but `tape.mojo` imports from `backward_ops.mojo` (backward functions).

**Solution**: Import order in Mojo handles this because:
- `tape.mojo` defines TapeNode, VariableRegistry first
- `backward_ops.mojo` imports those types
- `tape.mojo` then imports backward functions
- No actual circular dependency in type definitions

**Alternative if issues arise**: Move shared types to a separate `types.mojo`:

```
shared/autograd/
├── types.mojo (TapeNode, SavedTensors, VariableRegistry)
├── tape.mojo (GradientTape, imports from types and backward_ops)
└── backward_ops.mojo (backward functions, imports from types)
```

---

### Task 5: Create Tests for Extracted Backward Operations

**File**: `/home/mvillmow/ml-odyssey/tests/shared/autograd/test_backward_ops.mojo`

```mojo
"""Tests for backward operation implementations.

Verifies that extracted backward functions produce correct gradients
when called through the GradientTape API.
"""

from testing import assert_true
from tests.shared.conftest import assert_almost_equal, assert_equal_int
from shared.core.extensor import ExTensor, zeros, ones, ones_like
from shared.autograd import Variable, GradientTape
from shared.autograd import variable_add, variable_multiply, variable_sum

# ============================================================================
# Binary Arithmetic Tests
# ============================================================================

fn test_backward_add_simple() raises:
    """Test add backward with simple tensors."""
    var tape = GradientTape()
    tape.enable()
    
    var shape: List[Int] = [2]
    var x_data = ones(shape, DType.float32)
    var y_data = ones(shape, DType.float32)
    
    var x = Variable(x_data, True, tape)
    var y = Variable(y_data, True, tape)
    
    var z = variable_add(x, y, tape)
    var loss = variable_sum(z, tape)
    loss.backward(tape)
    
    # Gradient should be ones for both inputs
    var grad_x = tape.get_grad(x.id)
    var grad_y = tape.get_grad(y.id)
    assert_equal_int(grad_x.numel(), 2)
    assert_almost_equal(grad_x._data.bitcast[Float32]()[0], 1.0, 1e-5)

fn test_backward_multiply_chain() raises:
    """Test multiply backward in a chain."""
    # y = x * x, loss = sum(y)
    # d_loss/d_x = 2*x
    ...

fn test_backward_relu_zero_crossing() raises:
    """Test ReLU backward at zero crossing."""
    ...

fn test_backward_sigmoid_output() raises:
    """Test sigmoid backward uses output correctly."""
    ...

fn test_backward_matmul_gradient() raises:
    """Test matmul backward produces correct gradients."""
    ...

fn test_backward_sum_reduction() raises:
    """Test sum backward expands gradient."""
    ...

fn test_backward_mean_scaling() raises:
    """Test mean backward scales gradient correctly."""
    ...

fn main() raises:
    test_backward_add_simple()
    test_backward_multiply_chain()
    test_backward_relu_zero_crossing()
    test_backward_sigmoid_output()
    test_backward_matmul_gradient()
    test_backward_sum_reduction()
    test_backward_mean_scaling()
    print("All backward_ops tests passed!")
```

---

## 4. Files to Modify/Create

### Files to Create

| File | Description | Est. Lines |
|------|-------------|------------|
| `shared/autograd/backward_ops.mojo` | Extracted backward implementations | ~200 |
| `tests/shared/autograd/test_backward_ops.mojo` | Tests for backward ops | ~150 |

### Files to Modify

| File | Changes |
|------|---------|
| `shared/autograd/tape.mojo` | Remove methods (lines 611-730), add import, update dispatch |

### Post-Extraction File Sizes

| File | Before | After | Change |
|------|--------|-------|--------|
| `tape.mojo` | 792 lines | ~650 lines | -142 lines |
| `backward_ops.mojo` | 30 lines | ~200 lines | +170 lines |

---

## 5. Testing Approach

### Test Commands

```bash
# Format check
pixi run mojo format shared/autograd/backward_ops.mojo

# Compile check  
pixi run mojo build shared/autograd/

# New unit tests
pixi run mojo test tests/shared/autograd/test_backward_ops.mojo

# Existing integration tests
pixi run mojo test tests/shared/core/test_variable_backward.mojo
pixi run mojo test tests/shared/autograd/test_sgd_momentum.mojo

# All autograd tests
pixi run mojo test tests/shared/autograd/

# Pre-commit hooks
pre-commit run --all-files
```

### Regression Tests

The following existing tests validate the backward operations work correctly:
- `tests/shared/core/test_variable_backward.mojo` - Variable.backward() API
- `tests/shared/autograd/test_sgd_momentum.mojo` - Optimizer with gradients
- `tests/shared/core/test_arithmetic_backward.mojo` - Core arithmetic backward

---

## 6. Success Criteria

### Functional Requirements

- [ ] All 10 backward operations extracted to `backward_ops.mojo`
- [ ] `tape.mojo` reduced to ~650 lines (from 792)
- [ ] All existing tests pass without modification
- [ ] No circular import errors
- [ ] GradientTape.backward() works identically to before

### Quality Requirements

- [ ] Clean separation of concerns (tape logic vs backward ops)
- [ ] Comprehensive docstrings on all backward functions
- [ ] No warnings from `mojo format`
- [ ] Pre-commit hooks pass

### Documentation Requirements

- [ ] Module docstring explains design rationale
- [ ] Each backward function documented with Args/Raises sections

---

## 7. Risk Assessment

### Low Risk

- **API compatibility**: No public API changes - internal refactoring only
- **Test coverage**: Existing tests catch regressions

### Medium Risk

- **Circular imports**: Mitigated by passing components instead of GradientTape
- **Mojo import order**: May need to adjust if import issues arise

### Deferred (Out of Scope)

- **Dispatch table (#2586)**: Requires Dict support in Mojo stdlib
- **Additional operations**: OP_POWER, OP_SOFTMAX, etc. not yet implemented

---

## 8. Implementation Order

1. **Phase 1**: Create `backward_ops.mojo` with all 10 backward functions
2. **Phase 2**: Update `tape.mojo` imports and dispatch 
3. **Phase 3**: Remove old `_backward_*_by_idx` methods from GradientTape
4. **Phase 4**: Run all tests and fix any issues
5. **Phase 5**: Create test file for backward ops
6. **Phase 6**: Create PR linked to issue #2633

---

## Resource Usage

- **Tool calls used**: 15/50
- **Steps used**: 15/50
